libsolidity/analysis/FunctionCallGraph.h

```diff
@@ -0,0 +1,94 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+
+#include <vector>
+
+namespace solidity::frontend
+{
+
+class FunctionCallGraphBuilder: private ASTConstVisitor
+{
+public:
+	struct Node;
+
+	struct NodeCompare
+	{
+		// this member is required to let container be aware that
+		// comparator is capable of dealing with types other than key
+		using is_transparent = std::true_type;
+
+		bool operator()(std::shared_ptr<Node> const _lhs, std::shared_ptr<Node> const _rhs) const
+		{
+			return _lhs->callable < _rhs->callable;
+		}
+		bool operator()(std::shared_ptr<Node> const _lhs, CallableDeclaration const* _rhs) const
+		{
+			return _lhs->callable < _rhs;
+		}
+		bool operator()(CallableDeclaration const* _lhs, std::shared_ptr<Node> const _rhs) const
+		{
+			return _lhs < _rhs->callable;
+		}
+	};
+
+	typedef std::set<std::shared_ptr<Node>, NodeCompare> NodeSet;
+
+	struct Node
+	{
+		Node(CallableDeclaration const* _callable): callable(_callable) {}
+		// Definition of this function/modifier/event, might be nullptr for
+		// implicit constructors
+		CallableDeclaration const* callable = nullptr;
+
+		// Calls that this function/modifier/event makes to other callables
+		NodeSet calls;
+	};
+
+	struct ContractCallGraph
+	{
+		ContractCallGraph(ContractDefinition const& _contract) :contract(_contract) {}
+
+		/// Contract for which this is the graph
+		ContractDefinition const& contract;
+		/// Calls made at creation time (constructor, state variables, ..)
+		NodeSet creationCalls;
```

Comment:
Since the inner calls  should be the same at creation and at deploy time, you could use a`std::map<ASTNode const*, ASTNode const*>` for the edges of the graph and just have a `std::vector<ASTNode const*>` for the entry points at deploy time and a sigle node containing the constructor for the creation time.

Reply:
It made things a bit easier when comparing the generated functions when I had it split like this as I could simply directly compare the `creationCalls` and `runtimeCalls` sets with the generated functions...

Reply:
I had a single note at first but noticed that to do the actual comparison it is easier to have to separate sets. I suppose I could also manually create those sets when I am doing the comparison, going through the calls from the c'tor and adding each recursively..

Reply:
Looking towards other use-cases, I think having a proper graph and then extracting the set of reachable functions from it as a second stop could be cleaner.

Reply:
Alright

Reply:
One problem I see with the `map` approach: Some constructors are implicit, which means there is no `ASTNode` for them, which means that the initialization of state variables which I'd usually model as calls from the c'tor can't be done that way..

Reply:
Ah, I suppose I can use the contractdefinition instead of the functiondefinition as "constructor"